plugins/graph_algorithm/src/algorithms/neighborhood.cpp

```diff
@@ -0,0 +1,142 @@
+#include "graph_algorithm/algorithms/neighborhood.h"
+
+namespace hal
+{
+    class Gate;
+
+    namespace graph_algorithm
+    {
+        Result<std::vector<std::vector<u32>>> get_neighborhood(NetlistGraph* graph, const std::vector<Gate*>& start_gates, u32 order, NetlistGraph::Direction direction, u32 min_dist)
+        {
+            if (!graph)
+            {
+                return ERR("graph is a nullptr");
+            }
+
+            if (start_gates.empty())
+            {
+                return ERR("no start gates provided");
+            }
+
+            igraph_vector_int_t i_gates;
+            if (auto res = graph->get_vertices_from_gates_igraph(start_gates); res.is_ok())
+            {
+                i_gates = std::move(res.get());
+            }
+            else
+            {
+                return ERR(res.get_error());
+            }
+
+            auto res = get_neighborhood_igraph(graph, &i_gates, order, direction, min_dist);
+
+            igraph_vector_int_destroy(&i_gates);
+
+            if (res.is_error())
+            {
+                return ERR(res.get_error());
+            }
+
+            return res;
+        }
+
+        Result<std::vector<std::vector<u32>>> get_neighborhood(NetlistGraph* graph, const std::vector<u32>& start_vertices, u32 order, NetlistGraph::Direction direction, u32 min_dist)
+        {
+            if (!graph)
+            {
+                return ERR("graph is a nullptr");
+            }
+
+            if (start_vertices.empty())
+            {
+                return ERR("no start vertices provided");
+            }
+
+            igraph_vector_int_t i_gates;
+            if (auto res = igraph_vector_int_init(&i_gates, start_vertices.size()); res != IGRAPH_SUCCESS)
+            {
+                return ERR(igraph_strerror(res));
+            }
+
+            for (u32 i = 0; i < start_vertices.size(); i++)
+            {
+                VECTOR(i_gates)[i] = start_vertices.at(i);
+            }
+
+            auto res = get_neighborhood_igraph(graph, &i_gates, order, direction, min_dist);
+
+            igraph_vector_int_destroy(&i_gates);
+
+            if (res.is_error())
+            {
+                return ERR(res.get_error());
+            }
+
+            return res;
+        }
+
+        Result<std::vector<std::vector<u32>>> get_neighborhood_igraph(NetlistGraph* graph, const igraph_vector_int_t* start_gates, u32 order, NetlistGraph::Direction direction, u32 min_dist)
+        {
+            if (!graph)
+            {
+                return ERR("graph is a nullptr");
+            }
+
+            igraph_vs_t v_sel;
+            if (auto res = igraph_vs_vector(&v_sel, start_gates); res != IGRAPH_SUCCESS)
+            {
+                return ERR(igraph_strerror(res));
+            }
```

Comment:
Here's a tip: many of the `vs` functions have an "immediate" version with `vss` in the name. Instead of `igraph_vs_vector()`, you could use `igraph_vss_vector()`, which _returns_ a vertex selector instead of an error code. `vss` functions cannot fail, no need to check for errors. Using the `vss` version often simplifies code considerably.